### PR TITLE
Fix Material 3 `AppBar.leading` action IconButtons

### LINKED ISSUE #141361: `AppBar` back button focus/hover circle should not fill up whole height
### Steps to reproduce

1. Set up two screens with an `AppBar` and add a way to navigate to the other one
2. Run the app. (I tried macOS, CanvasKit, and skwasm)
3. Navigate to the second page
4. Hover over the automatic back button in the top app bar
5. Notice its hover circle fills up the whole height rather than have a consistent hover target

### Expected results

The circle from hovering over the back button wouldn't fill up the entire height of the app bar, nor expand when the height increases.

### Actual results

The back button hover circle fills up the entire height of the app bar, both with the default height and at increased heights. See screenshots below for reference.

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: const HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Home screen'),
        elevation: 8,
        toolbarHeight: 96,
      ),
      body: Center(
        child: TextButton(
          child: const Text('Go to other screen'),
          onPressed: () {
            Navigator.push(
              context,
              MaterialPageRoute(
                builder: (BuildContext context) => const SecondScreen(),
              ),
            );
          },
        ),
      ),
    );
  }
}

class SecondScreen extends StatelessWidget {
  const SecondScreen({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Other screen'),
        elevation: 8,
        toolbarHeight: 96,
      ),
      body: const Center(
        child: Text(
          'Other screen content',
        ),
      ),
    );
  }
}
```

</details>

### Screenshots

<details open>
<summary>Screenshots</summary>

Fill up whole height (default height):
Example of back button expanding to default appbar height

Fill up whole height (expanded height):
Example of back button expanding to full appbar height

Expected behavior:
Expected shape of hover circle

</details>

### Flutter version info

<details><summary>Version</summary>

```console
Flutter 3.19.0-3.0.pre.43 • channel main • https://github.com/flutter/flutter
Framework • revision f961fdf2ba (4 hours ago) • 2024-01-10 17:37:07 -0500
Engine • revision a045134c91
Tools • Dart 3.4.0 (build 3.4.0-14.0.dev) • DevTools 2.31.0
```

</details>

### LINKED ISSUE #154393: [Material 3] Date Range Picker close button has incorrect shape
### Steps to reproduce

Run a minimal `showDateRangePicker` sample.

### Expected results

after

before_size

### Actual results

before

after_size

### Specs
The date range picker is missing the close button padding as seen in the specs.

Screenshot 2024-08-29 at 17 59 03

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: Builder(
            builder: (BuildContext context) => ElevatedButton(
              onPressed: () {
                showDateRangePicker(
                  context: context,
                  firstDate: DateTime.now(),
                  lastDate: DateTime.now().add(const Duration(days: 30)),
                );
              },
              child: const Text('Show Date Range Picker'),
            ),
          ),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[!] Flutter (Channel main, 3.25.0-1.0.pre.172, on macOS 15.0 24A5327a darwin-arm64, locale
    en-US)
    • Flutter version 3.25.0-1.0.pre.172 on channel main at /Users/tahatesser/Code/flutter
    ! Upstream repository git@github.com:TahaTesser/flutter.git is not the same as
      FLUTTER_GIT_URL
    • FLUTTER_GIT_URL = git@github.com:TahaTesser/flutter.gitexport
    • Framework revision 9d971fada1 (25 hours ago), 2024-08-29 10:02:23 +0300
    • Engine revision bad94641d2
    • Dart version 3.6.0 (build 3.6.0-193.0.dev)
    • DevTools version 2.39.0
    • If those were intentional, you can disregard the above warnings; however it is
      recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/tahatesser/Code/android-sdk
    • Platform android-35, build-tools 35.0.0
    • ANDROID_SDK_ROOT = /Users/tahatesser/Code/android-sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode-beta.app/Contents/Developer
    • Build 16A5230g
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)

[✓] VS Code (version 1.92.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.94.0

[✓] Connected device (4 available)
    • Taha’s iPhone (mobile)          • 00008120-001C559034DB401E • ios            • iOS 18.0
      22A5350a
    • macOS (desktop)                 • macos                     • darwin-arm64   • macOS
      15.0 24A5327a darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad     • darwin         • macOS
      15.0 24A5327a darwin-arm64
    • Chrome (web)                    • chrome                    • web-javascript • Google
      Chrome 128.0.6613.113

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.

```

</details>

### PR DESCRIPTION
**Pull request by TahaTesser**
Fixes [`AppBar` back button focus/hover circle should not fill up whole height](https://github.com/flutter/flutter/issues/141361)
Fixes [[Material 3] Date Range Picker close button has incorrect shape](https://github.com/flutter/flutter/issues/154393)

This updates the leading condition added in https://github.com/flutter/flutter/pull/110722

### Code sample

<details>
<summary>expand to view the code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: SingleChildScrollView(
          child: Column(
            children: [
              Column(
                spacing: 10.0,
                mainAxisSize: MainAxisSize.min,
                children: <Widget>[
                  AppBar(
                    leading: BackButton(
                      style: IconButton.styleFrom(backgroundColor: Colors.red),
                    ),
                    backgroundColor: Theme.of(context).colorScheme.secondaryContainer,
                    title: const Text('AppBar with BackButton'),
                  ),
                  AppBar(
                    leading: CloseButton(
                      style: IconButton.styleFrom(backgroundColor: Colors.red),
                    ),
                    backgroundColor: Theme.of(context).colorScheme.secondaryContainer,
                    title: const Text('AppBar with CloseButton'),
                  ),
                  AppBar(
                    leading: DrawerButton(
                      style: IconButton.styleFrom(backgroundColor: Colors.red),
                    ),
                    backgroundColor: Theme.of(context).colorScheme.secondaryContainer,
                    title: const Text('AppBar with DrawerButton'),
                  ),
                ],
              ),
              const Divider(),
              Column(
                spacing: 10.0,
                mainAxisSize: MainAxisSize.min,
                children: <Widget>[
                  AppBar(
                    leading: BackButton(
                      style: IconButton.styleFrom(backgroundColor: Colors.red),
                    ),
                    backgroundColor: Theme.of(context).colorScheme.secondaryContainer,
                    toolbarHeight: 100.0,
                    title: const Text('AppBar with custom height'),
                  ),
                  AppBar(
                    leading: CloseButton(
                      style: IconButton.styleFrom(backgroundColor: Colors.red),
                    ),
                    backgroundColor: Theme.of(context).colorScheme.secondaryContainer,
                    toolbarHeight: 100.0,
                    title: const Text('AppBar with custom height'),
                  ),
                  AppBar(
                    leading: DrawerButton(
                      style: IconButton.styleFrom(backgroundColor: Colors.red),
                    ),
                    backgroundColor: Theme.of(context).colorScheme.secondaryContainer,
                    toolbarHeight: 100.0,
                    title: const Text('AppBar with custom height'),
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

</details>

### Before

Screenshot 2024-09-04 at 12 38 05

### After

Screenshot 2024-09-04 at 12 38 28

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/app_bar.dart [hunk 1]
```diff
@@ -966,9 +966,14 @@ class _AppBarState extends State<AppBar> {
           );
         }
 
+        final bool isIconButton = leading is IconButton
+          || leading is CloseButton
+          || leading is BackButton
+          || leading is DrawerButton;
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/app_bar.dart:972]: Can we make `_ActionButton` extends `IconButton`? Seems `_ActionButton` just returns a `IconButton`

**Review by QuncCccccc — APPROVED** [review]
LGTM!

File: packages/flutter/lib/src/material/app_bar.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (TahaTesser) [reply packages/flutter/lib/src/material/app_bar.dart:972]: While exploring this suggestion I realized i also missed `EndDrawerButton` in this check.  This really goes to show this type of fix is not very robust.

**Comment by QuncCccccc** [comment]
> @QuncCccccc this seems to be failing Google testing

I'm working on it!😎

**Comment by TahaTesser** [comment]
@QuncCccccc thanks for helping!

**Pull request merged**